src/shadowbox/server/manager_service.ts

```diff
@@ -168,6 +173,46 @@ export class ShadowsocksManagerService {
     }
   }
 
+  public async setAccessKeyQuota(req: RequestType, res: ResponseType, next: restify.Next) {
+    try {
+      logging.debug(`setAccessKeyQuota request ${JSON.stringify(req.params)}`);
+      const accessKeyId = req.params.id;
+      const quota = req.params.quota;
+      if (!quota || !quota.quotaBytes || !quota.windowHours) {
+        return next(new restify.InvalidArgumentError(
+            'Must provide a quota value with "quotaBytes" and "windowHours"'));
+      }
+      if (quota.quotaBytes < 0 || quota.windowHours < 0) {
+        return next(new restify.InvalidArgumentError('Must provide positive quota values'));
+      }
+      const success = await this.accessKeys.setAccessKeyQuota(accessKeyId, quota);
+      if (!success) {
+        return next(new restify.NotFoundError(`No access key found with id ${accessKeyId}`));
```

Comment:
The method could potentially fail on invalid input, no?

Reply:
It could, hence I added (duplicate) input checks above to serve a 409. Repository methods can also fail due to I/O, but we currenlty have no way of differentiating. I believe we need to refactor the server repository to throw typed errors that allow the caller to take the appropriate action based on different failure conditions. 

I added a TODO to support typed errors in the access key repository in 2808f76.